gk/bpf.c

```diff
@@ -146,6 +147,50 @@ pkt_ctx_to_pkt(struct gk_bpf_pkt_ctx *ctx)
 	return frame->pkt;
 }
 
+static int
+update_pkt_priority(struct rte_mbuf *pkt, uint16_t proto, int priority,
+	struct gatekeeper_if *iface)
+{
+	uint32_t mask;
+	struct rte_ether_hdr *eth_hdr =
+		rte_pktmbuf_mtod(pkt, struct rte_ether_hdr *);
+	if (eth_hdr == NULL) {
+		G_LOG(ERR, "bpf: failed to cast the start of the data in mbuf to type (struct rte_ether_hdr *) at %s\n",
+			__func__);
+		return -1;
+	}
+
+	if (proto == RTE_ETHER_TYPE_IPV4) {
+		struct rte_ipv4_hdr *ip4hdr = pkt_out_skip_l2(iface, eth_hdr);
+
+		mask = (1 << 2) - 1;
+		/* Update the IP header. */
+		ip4hdr->type_of_service = (priority << 2) |
+			(ip4hdr->type_of_service & mask);
+
+		/*
+		 * The IP header checksum filed must be set to 0
+		 * in order to offload the checksum calculation.
+		 */
+		ip4hdr->hdr_checksum = 0;
+
+		pkt->l3_len = sizeof(struct rte_ipv4_hdr);
+		/* Offload checksum computation for the outer IPv4 header. */
+		pkt->ol_flags |= (PKT_TX_IPV4 | PKT_TX_IP_CKSUM);
```

Comment:
Since we are updating a single byte in the header, namely field `type_of_service`, can't we just decrement the old value and increment the new one as we do in `update_ip_hop_count()`?

Reply:
I don't think so, as in the ``update_ip_hop_count()`` we are updating the TTL by subtracting ``1`` from it. However, here we are updating the ``type_of_service``, the difference between the old and new values of this byte can be various values. So, we cannot simple use that.

Reply:
This [RFC1624](https://tools.ietf.org/html/rfc1624) may be useful here.

Reply:
I think we can use `[Eqn. 3]` of this RFC. It's not clear for me if `m` is `type_of_service` or `type_of_service << 8`, though. Could you test it?

Reply:
In fact, we can use `[Eqn. 4]`, which is a bit simpler.

Reply:
I think ``m`` depends on the byteorder of the host machine, we can use the host order value of the first two bytes to calculate the checksum instead.